### PR TITLE
Support timezone in format `'+XX:XX'`

### PR DESCRIPTION
**Pull request by whjpji**
As the discussions in #190, it is straight-forward to implement `'+XX:XX'` formatted timezones within the `TimeZoneId`, using the lower byte of `i16` to represent the offset in quarters.

### CODE REVIEW

**Comment by jinmingjian** [comment]
@whjpji tests_integ_date_time_functions failed locally

**Comment by jinmingjian** [comment]
@whjpji and I have found that the last merge PR has broken the original meta store. Because we change the meta of BqlType...

**Comment by whjpji** [comment]
> @whjpji and I have found that the last merge PR has broken the original meta store. Because we change the meta of BqlType...

@jinmingjian One way is to split `DateTime(Option<TimeZoneId>)` up to `DateTime` and `DateTimeTz(TimeZoneId)` (appended to the back). Should we make this change?

**Comment by whjpji** [comment]
@jinmingjian Further more, we might have breaking changes of the `BqlType` or other metadata in the future. Is there versioning support in the ser/de of the metadata?

File: crates/base/src/datetimes.rs [hunk 1]
```diff
@@ -49,13 +51,17 @@ pub struct BaseTimeZone {
...
+/// * for values `x` that `x >= 0`, the index of a variety of the `enum` `chrono_tz::Tz`. There
+///   are 594 variants in total, so the it needs 10 lower bits to store the index.
+/// * for values `x` that `x < 0`, its lower byte of it (an `i8`) represents the fixed time zone
+///   offset in quarters. E.g., `5` means `'+01:15'` and `-12` means `'-03:00'`, etc.
+///   Note, `'-00:00'` will be converted to `'+00:00'`.
 #[derive(
     Serialize, Deserialize, Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash,
 )]
-pub struct TimeZoneId(u16);
+pub struct TimeZoneId(i16);
```
Comment (jinmingjian) [inline crates/base/src/datetimes.rs:64]: I make a thought: we just use TimeZoneId(i32). The performance impact is rare, in fact. We just use offset in seconds. We can discard TimeZoneId as well. That depends on your thinking.

File: crates/base/src/datetimes.rs [hunk 2]
```diff
@@ -49,13 +51,17 @@ pub struct BaseTimeZone {
     offset: i32,
```
Comment (jinmingjian) [inline crates/base/src/datetimes.rs:51]: merge offset with TimeZoneId(i32)

File: crates/base/src/datetimes.rs [hunk 3]
```diff
@@ -71,15 +77,30 @@ impl FromStr for TimeZoneId {
     type Err = BaseError;
 
     fn from_str(tz_name: &str) -> BaseResult<TimeZoneId> {
```
Comment (jinmingjian) [inline crates/base/src/datetimes.rs:79]: for the from_str and further tz string included logic, it is better like this: all string is convert to offset in seconds. We do not maintain two internal id form and not any trick necessary. That's, we discard DST support although there is a room for future.

**Review by jinmingjian — COMMENTED** [review]
DateTime(Option<TimeZoneId>) -> DateTime(TimeZoneId). For any DateTime, it has a TimeZoneId. The default is UTC.

File: crates/base/src/datetimes.rs [hunk 4]
```diff
@@ -131,28 +152,60 @@ impl TimeZoneId {
...
         })
     }
 
-    // SAFETY: the timezone id only comes from a variant of `chrono_tz::Tz`, so it
-    // is always valid
+    /// Return the corresponding `chrono_tz::Tz` variant for the non-negative values.
+    ///
+    /// SAFETY: for non-negative values, the timezone id only comes from a variant of
+    /// `chrono_tz::Tz`, so it is always valid.
     fn tz(self) -> Tz {
```
Comment (jinmingjian) [inline crates/base/src/datetimes.rs:165]: unnecessary if we discard the id from Tz

File: crates/base/src/datetimes.rs [hunk 5]
```diff
@@ -131,28 +152,60 @@ impl TimeZoneId {
...
+    ///
+    /// * For negative values, its lower byte `i8` means the offset in quarters, will return the
+    ///   format as `"+XX:XX"`;
+    /// * For non-negative values, its lower byte 'i8' means a variant of `chrono_tz::Tz`, will
+    ///   return its name.
     pub fn name(self) -> &'static str {
-        self.tz().name()
+        if self.0 < 0 {
+            static NAMES: SyncLazy<Vec<String>> = SyncLazy::new(|| {
+                (-24 * 4 + 1..24 * 4)
```
Comment (jinmingjian) [inline crates/base/src/datetimes.rs:179]: just use offset in sec to discard such complex computing

File: crates/base/src/datetimes.rs (hunk 1 above)
Comment (jinmingjian) [reply crates/base/src/datetimes.rs:64]: pub struct TimeZoneId(i16) -> pub struct TimeZoneId(pub i32)
TimeZoneId(0) is just UTC. We do not need Tz::UTC -> TimeZoneId

**Review by jinmingjian — COMMENTED** [review]
I suggest a overhaul to simply our last tz-in-meta works.

**Comment by jinmingjian** [comment]
> @jinmingjian Further more, we might have breaking changes of the `BqlType` or other metadata in the future. Is there versioning support in the ser/de of the metadata?

Yes, but we have not now. You can propose one. We have version for whole TB. In the future, we may keep time base compatibility (i.e.,  two-year). 

Note: The end users hope you have no breaking change for ever.

**Comment by whjpji** [comment]
> DateTime(Option) -> DateTime(TimeZoneId). For any DateTime, it has a TimeZoneId. The default is UTC.

But CH's behavior is that `DateTime` without timezone is interpreted as the server's timezone. If the user left `DateTime` without timezone, he or she should expect different timezones in servers with different timezone settings.

More precisely, it is the case: assume at timezone `GMT-8` the user created a table and insert a datum into it,
```sql
create table test (a DateTime) ENGINE BaseStorage;
insert into test values ('2021-07-11 20:49:14');
select * from test;
```
The user will get `'2021-07-11 20:49:14'`. Then if the server is restarted as timezone `'UTC'`, and the user does another query. As a result, `'2021-07-11 12:49:14'` is returned.

I haven't tested this behavior in CH because I found it kind of difficult to setup a default storage engine for CH. Should we take such situation into consideration?

**Comment by jinmingjian** [comment]
> But CH's behavior is that `DateTime` without timezone is interpreted as the server's timezone. If the user left `DateTime` without timezone, he or she should expect different timezones in servers with different timezone settings.

My mistake, you're right. DateTime without tz has a default tz in current server. This is a dynamic value. So, we have two variants as your suggested above! (the current and appended new.)

**Comment by whjpji** [comment]
@jinmingjian Basically I agree with you to store only the offset in seconds of timezone is enough. We support `'Some State/Some City'` and `'+XX:XX'` formatted timezone as input, and support only `'+XX:XX'` formatted timezone as output, then this problem can be solved.

**Comment by jinmingjian** [comment]
> @jinmingjian Basically I agree with you to store only the offset in seconds of timezone is enough. We support `'Some State/Some City'` and `'+XX:XX'` formatted timezone as input, and support only `'+XX:XX'` formatted timezone as output, then this problem can be solved.

Yes, just have this enough.

**Comment by whjpji** [comment]
> Note: The end users hope you have no breaking change forever.

@jinmingjian Yes. But we may have a breaking change before our v1.0 released, to avoid breaking changes even later.

File: crates/base/src/datetimes.rs (hunk 5 above)
Comment (whjpji) [reply crates/base/src/datetimes.rs]: The name's calculation is still needed here. it is a little bit complicated but will be calculated only once.

**Comment by whjpji** [comment]
@jinmingjian I found that the `TimeZoneId(i32)` might enlarge the size of `BqlType` in memory. Is the disk store affected? 

If possible, how about storing the offset in minutes using `i16` instread of `i32`?

**Comment by jinmingjian** [comment]
> I found that the `TimeZoneId(i32)` might enlarge the size of `BqlType` in memory. Is the disk store affected?

we have a show_option_size! macro in base utils. The size of `BqlType` in memory is 8B, which is enough for 32bits. This is the alignment side effect of ```#[repr(C)]```.

**Comment by whjpji** [comment]
> we have a show_option_size! macro in base utils. The size of `BqlType` in memory is 8B, which is enough for 32bits. This is the alignment side effect of `#[repr(C)]`.

@jinmingjian I did [an experiment](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=0d6447cf027a05876d79289d5458a223), which shows that it is possible to compress the size of `BqlType` to 4 bytes (with `#[repr(C, u8)]`). It shows that the representation of tag value is different when using `#[repr(C)]` or not.

**Comment by jinmingjian** [comment]
> > we have a show_option_size! macro in base utils. The size of `BqlType` in memory is 8B, which is enough for 32bits. This is the alignment side effect of `#[repr(C)]`.
> 
> @jinmingjian I did [an experiment](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=0d6447cf027a05876d79289d5458a223), which shows that it is possible to compress the size of `BqlType` to 4 bytes (with `#[repr(C, u8)]`). It shows that the representation of tag value is different when using `#[repr(C)]` or not.

#[repr(C)] is wanted for stable layout because the default layout is UB-like. u8 is ok now, but if you have to add 256+ variants, what happened? Current 8B gives out sufficient margin.

**Pull request merged**